### PR TITLE
Cocotb dokümanı hem Türkçe hem İngilizce olacak şekilde eklendi.

### PR DESCRIPTION
**Pull request by yunusesergun**
Cocotb dokümanı hem Türkçe hem İngilizce olacak şekilde eklendi.

@alperyazar Review yapmak ister misin abi :D?

### CODE REVIEW

**Comment by alperyazar** [comment]
Ayıpsın :d

File: cocotb/README.md [hunk 1]
```diff
@@ -0,0 +1,312 @@
+# COCOTB (Coroutine Based Cosimulation Test Bench)
+
+Doğrulama, debug, test... FPGA tasarımında sadece ve sadece RTL kod yazıp çalışmasını beklemek çok zor. Yaptığımız projeler gereği tek seferde bir kodu çalıştırmamız imkansız değil fakat corner-case diyebileceğimiz koşullarda kodun çalışmasını sağlamak ayrı bir zorluk.
+
+Bunu başarabilmek için kart üzerinde gerçek zamanlı testler öncesi yapacağımız adımlar çok önemli: Kodu doğrulamak ve recursive olarak kodu düzeltmek. Buna da aslında RTL doğrulama diyebiliriz.
```
Comment (alperyazar) [inline cocotb/README.md]: recursive yerine iterative demek daha doğru bence.

File: cocotb/README.md [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+Doğrulama konusunda UVM gibi çokça kullanılan bir yöntem var ama bu yazıda Python temelli `cocotb`'den bahsedeceğim. Evet, verilog, system-verilog veya VHDL ile yazılan bir doğrulamadan bahsetmiyorum. Tamamiyle Python kullanarak RTL tasarımınızı özgürce doğrulayabiliyorsunuz. Ayrıca kullanımı çok kolay.
```
Comment (alperyazar) [inline cocotb/README.md]: özgürlüğü sağlayan şey ne? open source olması mı? Bir de cocotb ile UVM çelişen şeyler mi? Tam bilmediğimden soruyorum yani cocotb içeisinde UVM yapılamıyor mu, bir metodoloji değil mi sonuçta? Mesela: https://github.com/pyuvm/pyuvm

File: cocotb/README.md [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+Bu soruya verilebilecek birçok cevap var fakat en önemlilerini aşağıda
+sıraladım:
+
+- Python dili temel manada kullanım açısından öğrenmesi ve yazması kolay bir dil.
+- `cocotb` ileri düzeyde Python bilgisi gerektirmiyor.
+- Python'ın sağladığı çok geniş bir kütüphane içeriği var (numpy, pandas...
+  gibi). RTL tabanlı bir doğrulama aracında kolayca yapamayacağınız işlemleri
+  (rastgele sayı üretme, array yapıları, ethernet paketleri... vs.) Python'da çok kolay bir şekilde
+  yapabilirsiniz.
+- Jenkins entegrasyonu sağlaması adına built-in desteğine sahip yani git tabanlı
```
Comment (alperyazar) [inline cocotb/README.md]: Jenkins şart değil, komut satırından çalıştırabildiğin için Gitlab CI'da da çalışır.

File: cocotb/README.md [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+![cocotb-layers](./assets/cocotb_overview.svg)
+
+## COCOTB Kurulumu
+
+Cocotb'yi Windows10'da çalıştırdığım için kurulum rehberini Windows için anlatacağım. Direkt olarak Windows'u kullanmak bana mantıklı gelmedi çünkü dependency açısından problemler oluşturabiliyor. Kişisel olarak bu tarz konularda çok fazla problem yaşadığım için `Docker Desktop` kullandım. `Docker`'ı en ince ayrıntısıyla öğrenmeme gerek kalmadan kullanabildim. Cocotb'yi kurabilmek için Docker'ın ne olduğunu bilmek yeterli. Docker en basit tanımıyla temelinde bir sanallaştırma yöntemi. Daha fazla bilgi için internette çok fazla kaynak bulunmaktadır fakat şimdilik o kısımlara değinmeyeceğim.
+
+[Bu linkten](https://docs.docker.com/desktop/install/windows-install/) Docker Desktop indirilir. Sonrasında bilgisayara kurulur. Ben ayarları önerilende kullandım ve kurulumu bu şekilde yaptım. Sonrasında aşağıdaki komutu cmd ile çalıştırdım:
+
+```bash
+docker run -it -v C:\Users\yunus\Desktop\cocotb_deneme\:/home/example ubuntu:22.04
```
Comment (alperyazar) [inline cocotb/README.md]: Örneklerde `yunus` gibi yerler kalmasa daha güzel olabilir. `<WORKING DIRECTORY>` gibi placeholder isimler kullanabilirsin bence.

File: cocotb/README.md [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+```
+
+Yukarıdaki komut ile ubuntu 22.04 container'ını kullandım yani sanal olarak ubuntu 22.04'ü Windows10 işletim sistemli bilgisayarımda kullanabilir hale geldim. Ayrıca cocotb'yi çalışacağım bir Windows dizinini ubuntu container içerisinde bir dizine bağladım. Bu şekilde, Windows üzerinden yapacağım değişiklikler docker container içerisinde de yapılır.
```
Comment (alperyazar) [inline cocotb/README.md]: Sondaki yapılır değil de görünür olur dense bence daha doğru.

File: cocotb/README.md [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+![capture1](./assets/capture1.png)
+
+Eğer cmd'yi kapadıktan sonra tekrar aynı docker container'ına bağlanmak istiyorsak aşağıdaki komutları sırasıyla başka bir cmd'de kullanabiliriz:
+
+```bash
+docker start "id"
+docker attach "id"
+```
+
+`"id"` kısmına `Docker Desktop` uygulamasından bakılabilir. Docker Desktop uygulaması açılır, `Containers` kısmına girilir ve aşağıda mavi ok ile işaretlenen kısımdan ilgili container id kopyalanır.
```
Comment (alperyazar) [inline cocotb/README.md]: Burada genelde `"id"` değil `<id>` kullanılır.

File: cocotb/README.md [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+
+- verilog/vhdl kaynak dosyaları
+- Python cocotb kodu
+- Makefile
+
+Kaynak dosyaları ve Python kodunun ne işe yaradığı anlaşılıyor fakat Makefile ne işe yarıyor? Makefile, simülatör ile Python cocotb kodu ve kaynak dosyaları arasında bir köprü görevi görüyor diyebilirim. İçerisinde kaynak dosyaları, Python kodu, top modül, simülatör ismi vs. gibi bilgiler yer alıyor.
+
+İlk deneme için oluşturduğum projeyi ve dosyaları kısaca anlatayım.
+
+- counter_deneme.v: Reset sonrası input clock'a göre output counter register sürekli artar.
```
Comment (alperyazar) [inline cocotb/README.md]: `counter_deneme.v` gibi backtick iine alabilirsn.

File: cocotb/README.md [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+
+Doğrulama yaparken arka planda waveform oluşmasını sağlayabiliyoruz. Bunun için tek yapmamız gereken `make` ile `WAVES=1` kullanmak:
+
+```bash
+make WAVES=1
+```
+
+Komutu çalıştırdıktan sonra waveform, `Makefile` dosyasının yer aldığı dizinde `sim_build` dosyasının içerisinde `.fst` uzantısıyla oluşur.
+
+Kolayca görüntüleyebilmek için `Surfer` kullandım çünkü online hizmeti de var. Bunun haricinde `GTKWave` de kullanılabilir. [Bu linkten](https://surfer-project.org) online bir şekilde waveform görüntüleyebiliyoruz.
```
Comment (alperyazar) [inline cocotb/README.md:297]: Surfer iyimiş

**Review by alperyazar — COMMENTED** [review]
Türkçe olanı okudum, eline sağlık :)

**Comment by yunusesergun** [comment]
Teşekkürler yorumların için. Eve geçince gerekli güncellemeleri yapacağım.

File: cocotb/README.md (hunk 8 above)
Comment (yunusesergun) [reply cocotb/README.md:297]: Evet, birkaç özellik bakımından GTKwave'den daha iyi.

File: cocotb/README.md (hunk 2 above)
Comment (yunusesergun) [reply cocotb/README.md]: Haklısın. UVM bir metodoloji. Uyulması gereken kurallar oluyor. Benim kastettiğim daha çok SV+UVM bakış açısıydı. O şekilde güncelleyeyim dokümanın ilgili kısımlarını.

Bir de; özgürlükten kastım bir şeylere bağlı kalmadan doğrulama yapabilmek. Yani belli bir metodolojiye bağlı kalmıyoruz veya kurallar, dosyalar içinde kaybolmuyoruz. Onu kastetmeye çalışmıştım.

Comment (alperyazar) [reply cocotb/README.md]: Özgürlük kısmını anladım, bu her zaman iyi bir şey mi bilemedim 🤔

Comment (yunusesergun) [reply cocotb/README.md]: Belli bir metodolojiye bağlı kalma kısmında hak veriyorum aslında ya. Ben o paragrafı bir gözden geçireyim en iyisi. Durup dururken UVM'e sallamaya gerek yok :D

Comment (alperyazar) [reply cocotb/README.md]: 😬
